### PR TITLE
Add unfuse() to FusedRNNCell

### PR DESCRIPTION
**Pull request by szha**
@piiswrong This pr adds `unfuse()` to `FusedRNNCell` which generates an equivalent SequentialRNNCell of the FusedRNNCell of the same spec, which can be used for step unroll and can run on CPU. This helps solve the problem that users have to understand the naming convention of the FusedRNN parameters by taking care of the naming conventions inside the function.

### CODE REVIEW

**Comment by sergeykolychev** [comment]
@piiswrong 
@szha 
note: I applied this change in order to enable adam optimizer
```
diff --git a/example/rnn/cudnn_lstm_bucketing.py b/example/rnn/cudnn_lstm_bucketing.py
index 529b1fb..26683dc 100644
--- a/example/rnn/cudnn_lstm_bucketing.py
+++ b/example/rnn/cudnn_lstm_bucketing.py
@@ -124,7 +124,6 @@ def train(args):
         kvstore             = args.kv_store,
         optimizer           = args.optimizer,
         optimizer_params    = { 'learning_rate': args.lr,
-                                'momentum': args.mom,
                                 'wd': args.wd },
         initializer         = mx.init.Xavier(factor_type="in", magnitude=2.34),
         arg_params          = arg_params,
```
and running the rnn cudnn example I get perplexity all the way down to 1.25, just wondering, is this a valid result ?
```
:~/mxnet/example/rnn$ python cudnn_lstm_bucketing.py --gpus 0 --bidirectional 1 --optimizer adam
WARNING: discarded 89 sentences longer than the largest bucket.
WARNING: discarded 4 sentences longer than the largest bucket.
2017-03-13 16:39:55,219 Epoch[0] Batch [50]	Speed: 1050.17 samples/sec	Train-Perplexity=2185.947514
2017-03-13 16:39:56,901 Epoch[0] Batch [100]	Speed: 951.22 samples/sec	Train-Perplexity=138.512020
2017-03-13 16:39:58,420 Epoch[0] Batch [150]	Speed: 1053.24 samples/sec	Train-Perplexity=56.881415
2017-03-13 16:40:00,018 Epoch[0] Batch [200]	Speed: 1001.57 samples/sec	Train-Perplexity=37.439466
2017-03-13 16:40:01,665 Epoch[0] Batch [250]	Speed: 971.51 samples/sec	Train-Perplexity=23.562517
2017-03-13 16:40:03,269 Epoch[0] Batch [300]	Speed: 997.62 samples/sec	Train-Perplexity=15.977261
2017-03-13 16:40:04,869 Epoch[0] Batch [350]	Speed: 999.90 samples/sec	Train-Perplexity=10.845050
2017-03-13 16:40:06,358 Epoch[0] Batch [400]	Speed: 1074.82 samples/sec	Train-Perplexity=8.619930
2017-03-13 16:40:07,966 Epoch[0] Batch [450]	Speed: 995.42 samples/sec	Train-Perplexity=6.745537
2017-03-13 16:40:09,683 Epoch[0] Batch [500]	Speed: 931.65 samples/sec	Train-Perplexity=5.684115
2017-03-13 16:40:11,217 Epoch[0] Batch [550]	Speed: 1043.80 samples/sec	Train-Perplexity=4.806681
2017-03-13 16:40:12,784 Epoch[0] Batch [600]	Speed: 1020.87 samples/sec	Train-Perplexity=4.344307
2017-03-13 16:40:14,458 Epoch[0] Batch [650]	Speed: 955.70 samples/sec	Train-Perplexity=3.645527
2017-03-13 16:40:16,040 Epoch[0] Batch [700]	Speed: 1011.90 samples/sec	Train-Perplexity=3.142627
2017-03-13 16:40:17,622 Epoch[0] Batch [750]	Speed: 1011.31 samples/sec	Train-Perplexity=3.065108
2017-03-13 16:40:19,250 Epoch[0] Batch [800]	Speed: 982.96 samples/sec	Train-Perplexity=2.789179
2017-03-13 16:40:20,764 Epoch[0] Batch [850]	Speed: 1056.76 samples/sec	Train-Perplexity=2.607575
2017-03-13 16:40:22,448 Epoch[0] Batch [900]	Speed: 950.57 samples/sec	Train-Perplexity=2.402986
2017-03-13 16:40:23,965 Epoch[0] Batch [950]	Speed: 1054.40 samples/sec	Train-Perplexity=2.262135
2017-03-13 16:40:25,668 Epoch[0] Batch [1000]	Speed: 939.59 samples/sec	Train-Perplexity=2.142036
2017-03-13 16:40:27,270 Epoch[0] Batch [1050]	Speed: 999.06 samples/sec	Train-Perplexity=2.045724
2017-03-13 16:40:28,921 Epoch[0] Batch [1100]	Speed: 969.53 samples/sec	Train-Perplexity=1.958419
2017-03-13 16:40:30,593 Epoch[0] Batch [1150]	Speed: 956.76 samples/sec	Train-Perplexity=1.839736
2017-03-13 16:40:32,175 Epoch[0] Batch [1200]	Speed: 1011.36 samples/sec	Train-Perplexity=1.821883
2017-03-13 16:40:33,836 Epoch[0] Batch [1250]	Speed: 963.35 samples/sec	Train-Perplexity=1.692954
2017-03-13 16:40:35,427 Epoch[0] Batch [1300]	Speed: 1006.11 samples/sec	Train-Perplexity=1.667462
2017-03-13 16:40:35,641 Epoch[0] Train-Perplexity=1.778679
2017-03-13 16:40:35,641 Epoch[0] Time cost=42.337
2017-03-13 16:40:36,989 Epoch[0] Validation-Perplexity=1.571053
2017-03-13 16:40:38,747 Epoch[1] Batch [50]	Speed: 940.36 samples/sec	Train-Perplexity=1.477905
2017-03-13 16:40:40,406 Epoch[1] Batch [100]	Speed: 964.30 samples/sec	Train-Perplexity=1.457133
2017-03-13 16:40:42,027 Epoch[1] Batch [150]	Speed: 987.61 samples/sec	Train-Perplexity=1.433822
2017-03-13 16:40:43,747 Epoch[1] Batch [200]	Speed: 930.16 samples/sec	Train-Perplexity=1.406376
2017-03-13 16:40:45,370 Epoch[1] Batch [250]	Speed: 985.94 samples/sec	Train-Perplexity=1.392561
2017-03-13 16:40:46,957 Epoch[1] Batch [300]	Speed: 1008.47 samples/sec	Train-Perplexity=1.397298
2017-03-13 16:40:48,378 Epoch[1] Batch [350]	Speed: 1126.06 samples/sec	Train-Perplexity=1.419818
2017-03-13 16:40:49,922 Epoch[1] Batch [400]	Speed: 1036.79 samples/sec	Train-Perplexity=1.343282
2017-03-13 16:40:51,548 Epoch[1] Batch [450]	Speed: 983.80 samples/sec	Train-Perplexity=1.352846
2017-03-13 16:40:53,285 Epoch[1] Batch [500]	Speed: 921.34 samples/sec	Train-Perplexity=1.340008
2017-03-13 16:40:54,776 Epoch[1] Batch [550]	Speed: 1073.53 samples/sec	Train-Perplexity=1.331774
2017-03-13 16:40:56,406 Epoch[1] Batch [600]	Speed: 981.62 samples/sec	Train-Perplexity=1.321414
2017-03-13 16:40:57,937 Epoch[1] Batch [650]	Speed: 1045.22 samples/sec	Train-Perplexity=1.303707
2017-03-13 16:40:59,507 Epoch[1] Batch [700]	Speed: 1019.15 samples/sec	Train-Perplexity=1.317281
2017-03-13 16:41:01,134 Epoch[1] Batch [750]	Speed: 983.68 samples/sec	Train-Perplexity=1.285403
2017-03-13 16:41:02,804 Epoch[1] Batch [800]	Speed: 957.78 samples/sec	Train-Perplexity=1.274932
2017-03-13 16:41:04,417 Epoch[1] Batch [850]	Speed: 992.53 samples/sec	Train-Perplexity=1.275065
2017-03-13 16:41:06,104 Epoch[1] Batch [900]	Speed: 948.57 samples/sec	Train-Perplexity=1.279590
2017-03-13 16:41:07,641 Epoch[1] Batch [950]	Speed: 1040.61 samples/sec	Train-Perplexity=1.270453
2017-03-13 16:41:09,335 Epoch[1] Batch [1000]	Speed: 944.59 samples/sec	Train-Perplexity=1.258742
2017-03-13 16:41:10,929 Epoch[1] Batch [1050]	Speed: 1004.17 samples/sec	Train-Perplexity=1.250862
2017-03-13 16:41:12,623 Epoch[1] Batch [1100]	Speed: 944.90 samples/sec	Train-Perplexity=1.240166
2017-03-13 16:41:14,214 Epoch[1] Batch [1150]	Speed: 1005.52 samples/sec	Train-Perplexity=1.221636
2017-03-13 16:41:15,888 Epoch[1] Batch [1200]	Speed: 955.77 samples/sec	Train-Perplexity=1.214481
2017-03-13 16:41:17,358 Epoch[1] Batch [1250]	Speed: 1089.15 samples/sec	Train-Perplexity=1.215788
```

**Comment by szha** [comment]
@sergeykolychev sorry for the delayed reply, Sergey, and thanks for trying out. How does the test perplexity look like?

**Comment by sergeykolychev** [comment]
I run both stacked and non stacked tests for bidirectional version after second epoch
the validation data seem to confirm this very low perplexity, actually it reaches about 1.1 after few additional epochs

```
~/mxnet/example/rnn$ python cudnn_lstm_bucketing.py --gpus 0 --optimizer adam --stack-rnn 0 --bidirectional 1 --model-prefix binostack --test --load-epoch 2
WARNING: discarded 89 sentences longer than the largest bucket.
WARNING: discarded 4 sentences longer than the largest bucket.
2017-03-14 15:28:04,348 Epoch[0] Batch [5]	Speed: 509.28 samples/sec	Train-Perplexity=1.167252
2017-03-14 15:28:04,736 Epoch[0] Batch [10]	Speed: 413.11 samples/sec	Train-Perplexity=1.165158
2017-03-14 15:28:04,911 Epoch[0] Batch [15]	Speed: 915.17 samples/sec	Train-Perplexity=1.183640
2017-03-14 15:28:05,164 Epoch[0] Batch [20]	Speed: 633.00 samples/sec	Train-Perplexity=1.172807
2017-03-14 15:28:05,402 Epoch[0] Batch [25]	Speed: 672.44 samples/sec	Train-Perplexity=1.186968
2017-03-14 15:28:05,595 Epoch[0] Batch [30]	Speed: 830.00 samples/sec	Train-Perplexity=1.210309
2017-03-14 15:28:05,759 Epoch[0] Batch [35]	Speed: 974.04 samples/sec	Train-Perplexity=1.149023
2017-03-14 15:28:05,988 Epoch[0] Batch [40]	Speed: 698.33 samples/sec	Train-Perplexity=1.166088
2017-03-14 15:28:06,152 Epoch[0] Batch [45]	Speed: 980.14 samples/sec	Train-Perplexity=1.217705
2017-03-14 15:28:06,412 Epoch[0] Batch [50]	Speed: 614.95 samples/sec	Train-Perplexity=1.157740
2017-03-14 15:28:06,657 Epoch[0] Batch [55]	Speed: 654.41 samples/sec	Train-Perplexity=1.189062
2017-03-14 15:28:06,906 Epoch[0] Batch [60]	Speed: 641.91 samples/sec	Train-Perplexity=1.169281
2017-03-14 15:28:07,110 Epoch[0] Batch [65]	Speed: 783.94 samples/sec	Train-Perplexity=1.199286
2017-03-14 15:28:07,249 Epoch[0] Batch [70]	Speed: 1156.82 samples/sec	Train-Perplexity=1.183091
2017-03-14 15:28:07,483 Epoch[0] Batch [75]	Speed: 682.07 samples/sec	Train-Perplexity=1.157511
2017-03-14 15:28:07,711 Epoch[0] Batch [80]	Speed: 704.03 samples/sec	Train-Perplexity=1.210544
2017-03-14 15:28:07,939 Epoch[0] Batch [85]	Speed: 701.71 samples/sec	Train-Perplexity=1.183270
2017-03-14 15:28:08,171 Epoch[0] Batch [90]	Speed: 690.97 samples/sec	Train-Perplexity=1.188512
2017-03-14 15:28:08,358 Epoch[0] Batch [95]	Speed: 856.16 samples/sec	Train-Perplexity=1.199481
2017-03-14 15:28:08,554 Epoch[0] Batch [100]	Speed: 813.74 samples/sec	Train-Perplexity=1.239939
2017-03-14 15:28:08,816 Epoch[0] Batch [105]	Speed: 612.00 samples/sec	Train-Perplexity=1.177403
2017-03-14 15:28:08,965 Epoch[0] Batch [110]	Speed: 1076.90 samples/sec	Train-Perplexity=1.236693
~/mxnet/example/rnn$ python cudnn_lstm_bucketing.py --gpus 0 --optimizer adam --stack-rnn 1 --bidirectional 1 --model-prefix bistack --test --load-epoch 2
WARNING: discarded 89 sentences longer than the largest bucket.
WARNING: discarded 4 sentences longer than the largest bucket.
2017-03-14 15:28:27,072 Epoch[0] Batch [5]	Speed: 527.29 samples/sec	Train-Perplexity=1.212561
2017-03-14 15:28:27,459 Epoch[0] Batch [10]	Speed: 413.04 samples/sec	Train-Perplexity=1.223665
2017-03-14 15:28:27,742 Epoch[0] Batch [15]	Speed: 566.24 samples/sec	Train-Perplexity=1.224434
2017-03-14 15:28:27,955 Epoch[0] Batch [20]	Speed: 751.72 samples/sec	Train-Perplexity=1.211298
2017-03-14 15:28:28,202 Epoch[0] Batch [25]	Speed: 647.66 samples/sec	Train-Perplexity=1.186997
2017-03-14 15:28:28,580 Epoch[0] Batch [30]	Speed: 424.08 samples/sec	Train-Perplexity=1.248221
2017-03-14 15:28:28,757 Epoch[0] Batch [35]	Speed: 901.12 samples/sec	Train-Perplexity=1.235415
2017-03-14 15:28:28,952 Epoch[0] Batch [40]	Speed: 821.82 samples/sec	Train-Perplexity=1.259927
2017-03-14 15:28:29,163 Epoch[0] Batch [45]	Speed: 760.24 samples/sec	Train-Perplexity=1.227220
2017-03-14 15:28:29,409 Epoch[0] Batch [50]	Speed: 648.64 samples/sec	Train-Perplexity=1.216542
2017-03-14 15:28:29,692 Epoch[0] Batch [55]	Speed: 566.86 samples/sec	Train-Perplexity=1.219724
2017-03-14 15:28:29,905 Epoch[0] Batch [60]	Speed: 751.11 samples/sec	Train-Perplexity=1.217469
2017-03-14 15:28:30,152 Epoch[0] Batch [65]	Speed: 648.17 samples/sec	Train-Perplexity=1.241759
2017-03-14 15:28:30,415 Epoch[0] Batch [70]	Speed: 607.59 samples/sec	Train-Perplexity=1.258886
2017-03-14 15:28:30,663 Epoch[0] Batch [75]	Speed: 647.56 samples/sec	Train-Perplexity=1.180226
2017-03-14 15:28:30,858 Epoch[0] Batch [80]	Speed: 819.48 samples/sec	Train-Perplexity=1.274805
2017-03-14 15:28:31,070 Epoch[0] Batch [85]	Speed: 756.04 samples/sec	Train-Perplexity=1.249228
2017-03-14 15:28:31,248 Epoch[0] Batch [90]	Speed: 900.17 samples/sec	Train-Perplexity=1.272883
2017-03-14 15:28:31,461 Epoch[0] Batch [95]	Speed: 750.19 samples/sec	Train-Perplexity=1.209898
2017-03-14 15:28:31,673 Epoch[0] Batch [100]	Speed: 754.09 samples/sec	Train-Perplexity=1.244364
2017-03-14 15:28:31,919 Epoch[0] Batch [105]	Speed: 650.25 samples/sec	Train-Perplexity=1.242283
2017-03-14 15:28:32,149 Epoch[0] Batch [110]	Speed: 698.60 samples/sec	Train-Perplexity=1.256919
```

**Comment by loofahcus** [comment]
I think DropoutCell should be added when dropout of FusedRNNCell is non-zero.

**Comment by szha** [comment]
@loofahcus you're right. Sorry that I overlooked it. Will update.
